akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -242,27 +242,47 @@ private[stream] object Collect {
   override def initialAttributes: Attributes = DefaultAttributes.collect and SourceLocation.forLambda(pf)
 
   def createLogic(inheritedAttributes: Attributes) =
-    new SupervisedGraphStageLogic(inheritedAttributes, shape) with InHandler with OutHandler {
-
-      import Collect.NotApplied
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      private var buffer: OptionVal[Out] = OptionVal.none
 
-      val wrappedPf = () => pf.applyOrElse(grab(in), NotApplied)
+      override def preStart(): Unit = pull(in)
 
-      override def onPush(): Unit = withSupervision(wrappedPf) match {
-        case Some(result) =>
-          result match {
-            case NotApplied             => pull(in)
-            case result: Out @unchecked => push(out, result)
-            case _                      => throw new RuntimeException() // won't happen, compiler exhaustiveness check pleaser
+      def onPush(): Unit =
+        try {
+          val elem = grab(in)
+          import Collect.NotApplied
+          pf.applyOrElse(elem, NotApplied) match {
+            case NotApplied => pull(in)
+            case result: Out @unchecked =>
+              if (isAvailable(out)) {
+                push(out, result)
+                pull(in)
+              } else {
+                buffer = OptionVal.Some(result)
+              }
+            case _ => throw new RuntimeException() // won't happen, compiler exhaustiveness check pleaser
```

Comment:
```suggestion
            case _ => throw new IllegalStateException() // won't happen, compiler exhaustiveness check pleaser
```

Reply:
It doesn't really matter since it's not supposed to enter here, just code cleanup while you are at it. You can change in Filter at the same time.